Add support for systemd socket activation

abeltranmsft opened the pull request:
<!--

If you know who should review your pull request, please assign it to that
person, otherwise the pull request would get assigned randomly.

If your pull request is for a specific language, please add the appropriate
lang label.

-->

See https://github.com/grpc/grpc/pull/30485
This PR fixes issues with the unit test when running it multiple times in parallel

Vignesh2208 commented:
> @Vignesh2208 could you trigger the checks please?

Sorry I missed this because I wasn't assigned as reviewer and didn't get any notifications. I triggered the checks now and added myself as reviewer.

Vignesh2208 commented:
Looks like there a re a few minor issues (unused variables etc). Could you re-sync with master and re-upload.

abeltranmsft commented:
@Vignesh2208 I rebased, re-ran generate_projects.sh and fixed styling. Could you retrigger the checks please?

abeltranmsft commented:
@Vignesh2208 I have fixed the unused variable in the last commit. Could you retrigger the checks please?

abeltranmsft commented:
@Vignesh2208 I have made the suggested changes. Could you retrigger the checks please?

Vignesh2208 commented:
> @Vignesh2208 I have made the suggested changes. Could you retrigger the checks please?

Looks like sanity and clang-tidy is failing again unfortunately :(

abeltranmsft commented:
> Looks like sanity and clang-tidy is failing again unfortunately : I can't see the failures when clicking on details

@Vignesh2208 for some reason I can't see the failures when clicking the link. Could you paste them here please?

Vignesh2208 commented:
> > Looks like sanity and clang-tidy is failing again unfortunately : I can't see the failures when clicking on details
> 
> @Vignesh2208 for some reason I can't see the failures when clicking the link. Could you paste them here please?

Sanity Failure:

```
++ dirname tools/distrib/clang_format_code.sh
+ cd tools/distrib/../..
++ pwd
+ REPO_ROOT=/var/local/git/grpc
+ '[' true == '' ']'
+ export CLANG_FORMAT_ROOT=/var/local/git/grpc
+ CLANG_FORMAT_ROOT=/var/local/git/grpc
+ '[' '' '!=' '' ']'
+ tools/dockerfile/grpc_clang_format/clang_format_all_the_things.sh
--- /var/local/git/grpc/src/core/lib/iomgr/systemd_utils.cc	2022-12-20 17:31:39.347744338 +0000
+++ /tmp/tmp.9xtS41ZlBi	2022-12-20 17:34:49.451795443 +0000
@@ -33,8 +33,8 @@

 #ifdef HAVE_LIBSYSTEMD
 bool set_matching_sd_unix_fd(grpc_tcp_server* s,
-                             const grpc_resolved_address* addr, const int fd_start,
-                             const int n) {
+                             const grpc_resolved_address* addr,
+                             const int fd_start, const int n) {
   absl::StatusOr<std::string> addr_name = grpc_sockaddr_to_string(addr, true);
   for (int i = fd_start; i < fd_start + n; i++) {
     if (sd_is_socket_unix(i, SOCK_STREAM, 1, addr_name.value().c_str(), 0)) {
@@ -46,8 +46,9 @@
 }

 bool set_matching_sd_inet_fd(grpc_tcp_server* s,
-                             const grpc_resolved_address* addr, const int family,
-                             const int port, const int fd_start, const int n) {
+                             const grpc_resolved_address* addr,
+                             const int family, const int port,
+                             const int fd_start, const int n) {
   for (int i = fd_start; i < fd_start + n; i++) {
     int r_inet = sd_is_socket_inet(i, family, SOCK_STREAM, 1, (uint16_t)port);
     int r_addr = sd_is_socket_sockaddr(
--- /var/local/git/grpc/test/core/iomgr/tcp_server_posix_test.cc	2022-12-20 17:31:39.575744467 +0000
+++ /tmp/tmp.pAKQo8nJT1	2022-12-20 17:35:14.659800555 +0000
@@ -487,8 +487,7 @@
   const int enable = 1;
   setsockopt(pre_fd, SOL_SOCKET, SO_REUSEADDR, &enable, sizeof(int));

-  int b = bind(pre_fd,
-               reinterpret_cast<struct sockaddr*>(&address),
+  int b = bind(pre_fd, reinterpret_cast<struct sockaddr*>(&address),
                sizeof(address));
   if (b < 0) {
     gpr_log(GPR_ERROR, "Unable to bind inet socket: %m");
@@ -585,8 +584,7 @@
     return -1;
   }

-  int b = bind(pre_fd,
-               reinterpret_cast<struct sockaddr*>(&address),
+  int b = bind(pre_fd, reinterpret_cast<struct sockaddr*>(&address),
                sizeof(address));
   if (b < 0) {
     gpr_log(GPR_ERROR, "Unable to bind unix socket: %m");
```

Vignesh2208 commented:
The Cland-Tidy error:

```
/root/.cache/bazel/_bazel_root/d2dc70c3d9da3fab488ba0dcbbd35051/execroot/com_github_grpc_grpc/test/core/iomgr/tcp_server_posix_test.cc:628:14: error: use nullptr [modernize-use-nullptr,-warnings-as-errors]
  srand(time(0));
             ^
             nullptr
Suppressed 8203 warnings (8103 in non-user code, 63 NOLINT, 37 with check filters).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
1 warning treated as error
```

Vignesh2208 commented:
@abeltranmsft Apologies. Was OOO during the winter break. Pasted the errors above

abeltranmsft commented:
> @abeltranmsft Apologies. Was OOO during the winter break. Pasted the errors above

Thanks @Vignesh2208! I have rebased and fixed the sanity/tidy errors. Could you retrigger the checks please?

Vignesh2208 commented:
> > @abeltranmsft Apologies. Was OOO during the winter break. Pasted the errors above
> 
> Thanks @Vignesh2208! I have rebased and fixed the sanity/tidy errors. Could you retrigger the checks please?

Thanks!. Started the checks

drfloob commented on src/core/lib/iomgr/systemd_utils.cc:
```diff
@@ -0,0 +1,116 @@
+/*
```
nit: This should be a C++-style comment, each line starting with `//`. Given that we don't have a linter for this yet, I think we can fix it later. See https://github.com/grpc/grpc/pull/31923 for context.

Vignesh2208 commented:
Merging it. Let's hope it sticks this time.
